internal/service/opensearch/domain.go

```diff
@@ -0,0 +1,1177 @@
+						},
+						"custom_endpoint_certificate_arn": {
+							Type:             schema.TypeString,
+							Optional:         true,
+							ValidateFunc:     verify.ValidARN,
+							DiffSuppressFunc: isCustomEndpointDisabled,
+						},
+					},
+				},
+			},
+			"endpoint": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"kibana_endpoint": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"ebs_options": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"ebs_enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+						},
+						"iops": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"volume_size": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"volume_type": {
+							Type:         schema.TypeString,
+							Optional:     true,
+							Computed:     true,
+							ValidateFunc: validation.StringInSlice(opensearchservice.VolumeType_Values(), false),
+						},
+					},
+				},
+			},
+			"encrypt_at_rest": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+							ForceNew: true,
+						},
+						"kms_key_id": {
+							Type:             schema.TypeString,
+							Optional:         true,
+							Computed:         true,
+							ForceNew:         true,
+							DiffSuppressFunc: suppressEquivalentKmsKeyIds,
+						},
+					},
+				},
+			},
+			"node_to_node_encryption": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
+							Type:     schema.TypeBool,
+							Required: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+			"cluster_config": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Computed: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"dedicated_master_count": {
+							Type:             schema.TypeInt,
+							Optional:         true,
+							DiffSuppressFunc: isDedicatedMasterDisabled,
+						},
+						"dedicated_master_enabled": {
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  false,
+						},
+						"dedicated_master_type": {
+							Type:             schema.TypeString,
+							Optional:         true,
+							DiffSuppressFunc: isDedicatedMasterDisabled,
+						},
+						"instance_count": {
+							Type:     schema.TypeInt,
+							Optional: true,
+							Default:  1,
+						},
+						"instance_type": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  opensearchservice.OpenSearchPartitionInstanceTypeM3MediumSearch,
```

Comment:
wouldn't that mean a new version would need to be released whenever AWS adds support for a new instance type?

Reply:
It does, although we effectively get such updates for free due to the use of that `Values()` function. Updating the provider's version of the AWS SDK will be enough to pull in the updated list of instance types. Such API calls are very common across the various services in the codebase and mean that we don't have to maintain our own list of values and try to keep them in sync.

But, you're right, there's a tradeoff here; We can 1) Offer users the convenience of plan-time validation so they can avoid the potential of partially applied changes (because other resources may get applied before the OpenSearch domain receives an error back from the API - it does the same validation as we are doing here). or 2) Forego that plan-time validation and instead offer users the ability to specify any value at all there, typos included. The provider generally seems to favour 1. Given the ~ 1 week release cycles of the provider such an approach would seem reasonable to me.

Reply:
A quick drive-by note.
Yes, we get the up-to-date list of valid instance types by upgrading the AWS Go SDK dependency (and doing a release) but practitioners would then have to upgrade to the newer Terraform AWS Provider release.
Leaving no validation allows older provider versions to be able to specify a newer instance type that passes `plan` with a possible runtime error from the underlying AWS API during `apply`.
We don't constrain the EC2 Instance's `instance_type` argument, for example, even though the AWS Go SDK does provide a list of currently valid instance types.

Reply:
Ah, interesting. In addition to that note I've been doing some experimenting and it looks like the ES/OpenSearch APIs are sufficiently vague here that adding such validation is actually incorrect/too constraining.

For example, whilst `aws es list-elasticsearch-instance-types --elasticsearch-version 7.10` only shows `.elasticsearch` instance types, the console does (only!) enable you to specify a `.search` instance type and this behaviour goes back at least as far as the 5.5 version. The console's list of available versions changes depending on the version of ES/OpenSearch you select in accordance with https://docs.aws.amazon.com/opensearch-service/latest/developerguide/supported-instance-types.html.

So:

1. AWS Console (quite rightly) wants to force you to use the newer instance types
2. To get the TF provider to accurately reflect the elasticsearch_version->instance_type mapping, the `ValidateFunc` would somehow have to get access to the `elasticsearch_version` from the ResourceData which I don't think it can? So it
  * Will only be able to do a partial validation (it would prevent typos only; version+instance type mismatches would still be possible)
  * Will constrain the list of valid values in a backward-incompatible manner (would only provide `.elasticsearch` types)
  * Will need practitioners to upgrade to newer releases of the provider

I'll throw my other PR away now :smile:

Reply:
Nice! Will do.

Reply:
I think we're in agreement then that the validation should go. We can add something later if we decide that would be helpful.